CPP/DesktopToastsCppApp/DesktopNotificationManagerCompat.cpp

```diff
@@ -0,0 +1,287 @@
+
+        THROW_IF_FAILED(Module<OutOfProc>::GetModule().RegisterObjects());
+
+        s_registeredActivator = true;
+    }
+
+    void RegisterComServer(GUID clsid, const wchar_t exePath[])
+    {
+        // Turn the GUID into a string
+        OLECHAR* clsidOlechar;
+        StringFromCLSID(clsid, &clsidOlechar);
+        std::wstring clsidStr(clsidOlechar);
+        ::CoTaskMemFree(clsidOlechar);
+
+        // Create the subkey
+        // Something like SOFTWARE\Classes\CLSID\{23A5B06E-20BB-4E7E-A0AC-6982ED6A6041}\LocalServer32
+        std::wstring subKeyStr = LR"(SOFTWARE\Classes\CLSID\)" + clsidStr + LR"(\LocalServer32)";
+        LPCWSTR subKey = subKeyStr.c_str();
+
+        // Include -ToastActivated launch args on the exe
+        std::wstring exePathStr(exePath);
+        exePathStr = L"\"" + exePathStr + L"\" " + TOAST_ACTIVATED_LAUNCH_ARG;
+        exePath = exePathStr.c_str();
+
+        // We don't need to worry about overflow here as ::GetModuleFileName won't
+        // return anything bigger than the max file system path (much fewer than max of DWORD).
+        DWORD dataSize = static_cast<DWORD>((::wcslen(exePath) + 1) * sizeof(WCHAR));
+
+        // Register the EXE for the COM server
+        THROW_IF_FAILED(HRESULT_FROM_WIN32(::RegSetKeyValue(
+            HKEY_CURRENT_USER,
+            subKey,
+            nullptr,
+            REG_SZ,
+            reinterpret_cast<const BYTE*>(exePath),
+            dataSize)));
+    }
+
+    ComPtr<IToastNotifier> CreateToastNotifier()
+    {
+        EnsureRegistered();
+
+        ComPtr<IToastNotificationManagerStatics> toastStatics;
+        THROW_IF_FAILED(Windows::Foundation::GetActivationFactory(
+            HStringReference(RuntimeClass_Windows_UI_Notifications_ToastNotificationManager).Get(),
+            &toastStatics));
+
+        ComPtr<IToastNotifier> notifier;
+        if (s_aumid.empty())
+        {
+            THROW_IF_FAILED(toastStatics->CreateToastNotifier(&notifier));
+        }
+        else
+        {
+            THROW_IF_FAILED(toastStatics->CreateToastNotifierWithId(HStringReference(s_aumid.c_str()).Get(), &notifier));
+        }
+
+        return notifier;
+    }
+
+    ComPtr<IToastNotification> CreateToastNotification(IXmlDocument *content)
+    {
+        ComPtr<IToastNotificationFactory> factory;
+        THROW_IF_FAILED(Windows::Foundation::GetActivationFactory(
+            HStringReference(RuntimeClass_Windows_UI_Notifications_ToastNotification).Get(),
+            &factory));
+
+        ComPtr<IToastNotification> notification;
+        THROW_IF_FAILED(factory->CreateToastNotification(content, &notification));
+
+        return notification;
+    }
+
+    ComPtr<IXmlDocument> CreateXmlDocumentFromString(const wchar_t *xmlString)
+    {
+        ComPtr<IInspectable> docInspectable;
+        THROW_IF_FAILED(RoActivateInstance(HStringReference(RuntimeClass_Windows_Data_Xml_Dom_XmlDocument).Get(), docInspectable.ReleaseAndGetAddressOf()));
```

Comment:
does this need to be different? can you not use GetActivationFactory for this?

Reply:
Good catch, realized I can actually use MakeAndInitialize<> which is even cleaner